Modules/pyexpat.c

```diff
@@ -1631,101 +1627,152 @@ static int init_handler_descrs(void)
-    PyObject *errors_module;
-    PyObject *modelmod_name;
-    PyObject *model_module;
-    PyObject *tmpnum, *tmpstr;
-    PyObject *codes_dict;
-    PyObject *rev_codes_dict;
-    int res;
-    static struct PyExpat_CAPI capi;
-    PyObject *capi_object;
+    pyexpat_state *state = pyexpat_get_state(module);
+    Py_VISIT(state->xml_parse_type);
+    Py_VISIT(state->error);
+    return 0;
+}
 
-    if (errmod_name == NULL)
-        return NULL;
-    modelmod_name = PyUnicode_FromString(MODULE_NAME ".model");
-    if (modelmod_name == NULL)
-        return NULL;
+static int
+pyexpat_clear(PyObject *module)
+{
+    pyexpat_state *state = pyexpat_get_state(module);
+    Py_CLEAR(state->xml_parse_type);
+    Py_CLEAR(state->error);
+    return 0;
+}
 
-    if (PyType_Ready(&Xmlparsetype) < 0 || init_handler_descrs() < 0)
-        return NULL;
+static void
+pyexpat_free(void *module)
+{
+    pyexpat_clear((PyObject *)module);
+}
 
-    /* Create the module and add the functions */
-    m = PyModule_Create(&pyexpatmodule);
-    if (m == NULL)
-        return NULL;
+static int
+pyexpat_exec(PyObject *mod)
+{
+    PyObject *codes_dict = NULL, *rev_codes_dict = NULL;
+
+    pyexpat_state *state = pyexpat_get_state(mod);
+    state->xml_parse_type = (PyTypeObject *)PyType_FromModuleAndSpec(
+        mod, &_xml_parse_type_spec, NULL);
+    if (state->xml_parse_type == NULL) {
+        goto error;
+    }
+
+    if (init_handler_descrs(state) < 0) {
+        goto error;
+    }
 
     /* Add some symbolic constants to the module */
-    if (ErrorObject == NULL) {
-        ErrorObject = PyErr_NewException("xml.parsers.expat.ExpatError",
-                                         NULL, NULL);
-        if (ErrorObject == NULL)
-            return NULL;
+    state->error = PyErr_NewException("xml.parsers.expat.ExpatError",
+                                        NULL, NULL);
+    if (state->error == NULL) {
+        goto error;
+    }
+    
+    Py_INCREF(state->error);
+    if (PyModule_AddObject(mod, "error", state->error ) < 0) {
+        goto error;
+    }
+    Py_INCREF(state->error);
+    if (PyModule_AddObject(mod, "ExpatError", state->error ) < 0) {
+        goto error;
+    }
+    Py_INCREF(state->xml_parse_type);
+    if (PyModule_AddObject(mod, "XMLParserType",
+                           (PyObject *) state->xml_parse_type) < 0) {
+        goto error;
     }
-    Py_INCREF(ErrorObject);
-    PyModule_AddObject(m, "error", ErrorObject);
-    Py_INCREF(ErrorObject);
-    PyModule_AddObject(m, "ExpatError", ErrorObject);
-    Py_INCREF(&Xmlparsetype);
-    PyModule_AddObject(m, "XMLParserType", (PyObject *) &Xmlparsetype);
 
-    PyModule_AddStringConstant(m, "EXPAT_VERSION",
-                               XML_ExpatVersion());
+    if (PyModule_AddStringConstant(mod, "EXPAT_VERSION",
+                               XML_ExpatVersion()) < 0) {
+        goto error;
+    }
     {
         XML_Expat_Version info = XML_ExpatVersionInfo();
```

Comment:
You should store Py_BuildValue("(iii)", info.major, info.major, info.minor, info.micro) result into a temporary variable, to DECREF it on error.